clang/lib/Sema/SemaBounds.cpp

```diff
@@ -1875,55 +1875,95 @@ namespace {
+    // necessary side effects, and returns unknown bounds.
     //
-    // Check methods do the following for a node:
-    // 1. Recursively traverse all its children
-    // 2. Infer its (rvalue) bounds
-    // 3. Do work such as checking bounds declarations
-    //    or inserting bounds checks
-    BoundsExpr *TraverseStmt(Stmt *S, CheckedScopeSpecifier CSS) {
+    // The returned bounds expression may contain a modifying expression within
+    // it. It is the caller's responsibility to validate that the bounds
+    // expression is non-modifying.
+    //
+    // Check recursively checks the children of e and performs any
+    // necessary side effects on e.  Check and CheckLValue work together
+    // to traverse each expression in a CFG exactly once.
+    BoundsExpr *Check(Stmt *S, CheckedScopeSpecifier CSS) {
       if (!S)
         return CreateBoundsEmpty();
 
-      BoundsExpr *ResultBounds = CreateBoundsAlwaysUnknown();
+      if (Expr *E = dyn_cast<Expr>(S)) {
+        E = E->IgnoreParens();
+        S = E;
+        if (E->isLValue()) {
+          BoundsExpr *TargetBounds = nullptr;
+          CheckLValue(E, CSS, TargetBounds);
+          return CreateBoundsAlwaysUnknown();
+        }
+      }
 
-      if (Expr *E = dyn_cast<Expr>(S))
-        S = E->IgnoreParens();
+      BoundsExpr *ResultBounds = CreateBoundsAlwaysUnknown();
 
       switch (S->getStmtClass()) {
-        case Expr::UnaryOperatorClass: {
-          BoundsExpr *SubExprBounds = nullptr;
-          ResultBounds = CheckUnaryOperator(cast<UnaryOperator>(S),
-                                            CSS, SubExprBounds);
-          return AdjustRValueBounds(S, ResultBounds);
-        }
+        case Expr::UnaryOperatorClass:
+          ResultBounds = CheckUnaryOperator(cast<UnaryOperator>(S), CSS);
+          break;
         case Expr::CallExprClass:
           ResultBounds = CheckCallExpr(cast<CallExpr>(S), CSS);
-          return AdjustRValueBounds(S, ResultBounds);
-        case Expr::MemberExprClass:
-          ResultBounds = CheckMemberExpr(cast<MemberExpr>(S), CSS);
-          return AdjustRValueBounds(S, ResultBounds);
+          break;
         case Expr::ImplicitCastExprClass:
         case Expr::CStyleCastExprClass:
-        case Expr::BoundsCastExprClass: {
-          BoundsExpr *SubExprLValueBounds = nullptr;
-          ResultBounds = CheckCastExpr(cast<CastExpr>(S), CSS,
-                                       SubExprLValueBounds);
-          return AdjustRValueBounds(S, ResultBounds);
-        }
+        case Expr::BoundsCastExprClass:
+          ResultBounds = CheckCastExpr(cast<CastExpr>(S), CSS);
+          break;
         case Expr::BinaryOperatorClass:
         case Expr::CompoundAssignOperatorClass:
           ResultBounds = CheckBinaryOperator(cast<BinaryOperator>(S), CSS);
-          return AdjustRValueBounds(S, ResultBounds);
+          break;
         case Stmt::CompoundStmtClass: {
           CompoundStmt *CS = cast<CompoundStmt>(S);
```

Comment:
Following up from our discussion about checking the children of CompoundStmts:

Constructing the CFG "flattens" children of compound statements. For example, given this test code:
```
void compound_test(array_ptr<int> a : count(3), nt_array_ptr<char> c : count(4)) {
  checked {
    array_ptr<int> arr : count(3) = a;
    unchecked {
      arr = b; // error: declared bounds for arr are invalid after assignment
    }
  }
  unchecked {
    nt_array_ptr<char> buf : count(4) = c;
    buf = "ab"; // error: declared bounds for buf are invalid after assignment
  }
}
```
The following AST will be constructed as the body of the CFG:
```
CompoundStmt 0x1fac9ea5408
|-CompoundStmt 0x1fac9ea5110 _Checked checking-state bounds-and-types
| |-DeclStmt 0x1fac9ea5038
| | `-VarDecl 0x1fac9ea4f20  used arr '_Array_ptr<int>' cinit
| |   |-CountBoundsExpr 0x1fac9ea4fb0 'NULL TYPE' Element
| |   | `-IntegerLiteral 0x1fac9ea4f88 'int' 3
| |   `-ImplicitCastExpr 0x1fac9ea5008 '_Array_ptr<int>' <LValueToRValue>
| |     `-DeclRefExpr 0x1fac9ea4fe0 '_Array_ptr<int>' lvalue ParmVar 0x1fac9ea4ab8 'a' '_Array_ptr<int>'
| `-CompoundStmt 0x1fac9ea50f0 _Unchecked
|   `-BinaryOperator 0x1fac9ea50d0 '_Array_ptr<int>' '='
|     |-DeclRefExpr 0x1fac9ea5050 '_Array_ptr<int>' lvalue Var 0x1fac9ea4f20 'arr' '_Array_ptr<int>'
|     `-ImplicitCastExpr 0x1fac9ea50a0 '_Array_ptr<int>' <LValueToRValue>
|       `-DeclRefExpr 0x1fac9ea5078 '_Array_ptr<int>' lvalue ParmVar 0x1fac9ea4b58 'b' '_Array_ptr<int>'
`-CompoundStmt 0x1fac9ea5380 _Unchecked
  |-DeclStmt 0x1fac9ea5280
  | `-VarDecl 0x1fac9ea5168  used buf '_Nt_array_ptr<char>' cinit
  |   |-CountBoundsExpr 0x1fac9ea51f8 'NULL TYPE' Element
  |   | `-IntegerLiteral 0x1fac9ea51d0 'int' 4
  |   `-ImplicitCastExpr 0x1fac9ea5250 '_Nt_array_ptr<char>' <LValueToRValue>
  |     `-DeclRefExpr 0x1fac9ea5228 '_Nt_array_ptr<char>' lvalue ParmVar 0x1fac9ea4bf0 'c' '_Nt_array_ptr<char>'
  `-BinaryOperator 0x1fac9ea5360 '_Nt_array_ptr<char>' '='
    |-DeclRefExpr 0x1fac9ea5298 '_Nt_array_ptr<char>' lvalue Var 0x1fac9ea5168 'buf' '_Nt_array_ptr<char>'
    `-ImplicitCastExpr 0x1fac9ea5330 '_Nt_array_ptr<char>' <ArrayToPointerDecay> BoundsSafeInterface
      `-CHKCBindTemporaryExpr 0x1fac9ea5318 'char [3]' lvalue
        `-StringLiteral 0x1fac9ea52f8 'char [3]' lvalue "ab"
```
The following CFG will be constructed:
```
 [B2 (ENTRY)]
   Succs (1): B1

 [B1]
   1: _Array_ptr<int> arr : count(3) = a;
   2: arr = b
   3: _Nt_array_ptr<char> buf : count(4) = c;
   4: buf = "ab"
   Preds (1): B2
   Succs (1): B0

 [B0 (EXIT)]
   Preds (1): B1
```
Statements 1, 2, 3, and 4 from CFG block B1 will be traversed in `TraverseCFG`, so `Check` will be called on each of these four statements. `Check` will not be called on a CompoundStmt.